store_init.go

```diff
@@ -0,0 +1,278 @@
+		}
+		dbbs.instanceNamespace = "fil_instance_" + cfg.InstanceNamespace // everything is hard-coded to a `fil_` prefix
+	}
+
+	if cfg.ExtraPreloadNamespace != "" {
+		if !validNamespace.MatchString(cfg.ExtraPreloadNamespace) {
+			return nil, fmt.Errorf("provided instance-specific extra namespace '%s' does not match %s", cfg.ExtraPreloadNamespace, validNamespace.String())
+		}
+		dbbs.additionalPreloadNamespace = "fil_instance_" + cfg.ExtraPreloadNamespace
+	}
+
+	if cfg.CacheSizeGiB != 0 {
+		if cfg.CacheSizeGiB > 256 {
+			return nil, fmt.Errorf("requested cache size of %dGiB impossibly large", cfg.CacheSizeGiB)
+		}
+		dbbs.lruSizeBytes = int64(cfg.CacheSizeGiB << 30)
+	}
+
+	var dbConnCfg *pgxpool.Config
+	dbConnCfg, err = pgxpool.ParseConfig(cfg.PgxConnectString)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to parse connection string '%s': %w", cfg.PgxConnectString, err)
+	}
+	if !dbbs.isWritable {
+		dbConnCfg.ConnConfig.RuntimeParams["default_transaction_read_only"] = "TRUE"
+	}
+	if dbConnCfg.MaxConns > MaxPgxPoolSize {
+		dbConnCfg.MaxConns = MaxPgxPoolSize
+	}
+
+	var dbPool *pgxpool.Pool
+	dbPool, err = pgxpool.ConnectConfig(ctx, dbConnCfg)
+	defer func() {
+		if err != nil && dbPool != nil {
+			dbPool.Close()
+		}
+	}()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to connect to '%s': %w", cfg.PgxConnectString, err)
+	}
+	dbbs.dbPool = dbPool
+
+	var currentPgVersion int32
+	err = dbPool.QueryRow(ctx, `SELECT CURRENT_SETTING('server_version_num')::INTEGER`).Scan(&currentPgVersion)
+	if err != nil {
+		return nil, xerrors.Errorf("retrieving server version failed: %w", err)
+	}
+	if currentPgVersion < minimumPgVersion {
+		return nil, fmt.Errorf(
+			"code was tested on PostgreSQL version %d only: change the source if you want to try running on the older version %d",
+			minimumPgVersion, currentPgVersion,
+		)
+	}
+
+	dbSettingsExpected := "Encoding:SQL_ASCII Collate:C Ctype:C"
+	var dbSettings string
+	if err := dbPool.QueryRow(
+		ctx,
+		`
+		SELECT 'Encoding:' || pg_encoding_to_char(encoding) || ' Collate:' || datcollate || ' Ctype:' || datctype
+			FROM pg_database
+		WHERE datname = current_database()
+		`,
+	).Scan(&dbSettings); err != nil {
+		return nil, err
+	}
+	if dbSettings != dbSettingsExpected {
+		return nil, fmt.Errorf(
+			"unexpected database settings: you must create your database with something like `%s` for reliable and performant binary storage\n Current settings: %s\nExpected settings: %s",
+			`CREATE DATABASE {{name}} ENCODING='SQL_ASCII' LC_COLLATE='C' LC_CTYPE='C' TEMPLATE='template0'`,
+			dbSettings, dbSettingsExpected,
+		)
+	}
+
+	// check whether we are talking to a slave, and force RO to shut off tipset-tracking automatically
+	if dbbs.isWritable {
+		if _, err := dbPool.Exec(ctx, fmt.Sprintf(
+			"CREATE TEMPORARY TABLE %s ( pk INTEGER ) ON COMMIT DROP",
+			"tmptable_"+randBytesAsHex(),
+		)); err != nil {
```

Comment:
Check type of error here to avoid transient errors forcing RO mode.

Reply:
Good catch, fixed in https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/7ebc314e42a